bindings/experimental/distrdf/python/DistRDF/HeadNode.py

```diff
@@ -0,0 +1,377 @@
+import glob
+import logging
+import warnings
+
+import ROOT
+
+from DistRDF import Node
+
+logger = logging.getLogger(__name__)
+
+
+class FriendInfo(object):
+    """
+    A simple class to hold information about friend trees.
+
+    Attributes:
+        friend_names (list): A list with the names of the `ROOT.TTree` objects
+            which are friends of the main `ROOT.TTree`.
+
+        friend_file_names (list): A list with the paths to the files
+            corresponding to the trees in the `friend_names` attribute. Each
+            element of `friend_names` can correspond to multiple file names.
+    """
+
+    def __init__(self, friend_names=[], friend_file_names=[]):
+        """
+        Create an instance of FriendInfo
+
+        Args:
+            friend_names (list): A list containing the treenames of the friend
+                trees.
+
+            friend_file_names (list): A list containing the file names
+                corresponding to a given treename in friend_names. Each
+                treename can correspond to multiple file names.
+        """
+        self.friend_names = friend_names
+        self.friend_file_names = friend_file_names
+
+    def __bool__(self):
+        """
+        Define the behaviour of FriendInfo instance when boolean evaluated.
+        Both lists have to be non-empty in order to return True.
+
+        Returns:
+            bool: True if both lists are non-empty, False otherwise.
+        """
+        return bool(self.friend_names) and bool(self.friend_file_names)
+
+    def __nonzero__(self):
+        """
+        Python 2 dunder method for __bool__. Kept for compatibility.
+        """
+        return self.__bool__()
+
+
+class Factory(object):
+    """
+    A factory for different kinds of root nodes of the RDataFrame computation
+    graph, depending on the arguments to the RDataFrame constructor. Currently
+    can create a TreeHeadNode or an EntriesHeadNode.
+    """
+
+    @staticmethod
+    def get_headnode(*args):
+        """
+        Parses the arguments and compares them against the possible RDataFrame
+        constructors
+        """
+
+        # Early check that arguments are accepted by RDataFrame
+        ROOT.RDataFrame(*args)
+
+        firstarg = args[0]
+        if isinstance(firstarg, int):
```

Comment:
The lack of function overloading makes us do this kind of things :disappointed: I guess when RNtuple comes the RDataFrame constructor will accept an RNtuple (so there won't be any ambiguity here)

Reply:
We can probably provide a similar factory function `ROOT.RDF.Distributed.MakeNTupleDataFrame` and from that we can dispatch to a headnode type like `RNTupleHeadNode`. This would be in the spirit of needing minimal changes in user code that already uses `MakeNTupleDataFrame`